### PR TITLE
[SYCL] Remove arbitrary upper bound for attribute argument

### PR DESCRIPTION
**Pull request by elizabethandrews**
Remove arbitrary upper bounds for the following attributes -  

SYCLIntelSchedulerTargetFmaxMhz 
SYCLIntelLoopFuse
IntelFPGAPrivateCopies
IntelFPGABankBits
IntelFPGABankWidth
IntelFPGANumBanks
IntelFPGAMaxReplicates

Closes https://github.com/intel/llvm/issues/2911

### CODE REVIEW

**Comment by elizabethandrews** [comment]
@AaronBallman could you take a look at this to let me know your thoughts. I renamed the 'IntelSYCL' functions to be more generic so we can reuse it for other FPGA attributes as well. I think the function can be refactored but I thought I would upload this now just to get any feedback you may have before I work on the remaining attributes.

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -10094,8 +10094,8 @@ class Sema final {
   void AddIntelFPGABankBitsAttr(Decl *D, const AttributeCommonInfo &CI,
                                 Expr **Exprs, unsigned Size);
   template <typename AttrType>
-  void addIntelSYCLSingleArgFunctionAttr(Decl *D, const AttributeCommonInfo &CI,
-                                         Expr *E);
+  void addSingleArgFunctionAttr(Decl *D, const AttributeCommonInfo &CI,
```
Comment (AaronBallman) [inline clang/include/clang/Sema/Sema.h:10097]: I think it's actually a bit dangerous to rename this function -- it's not a generic function that can be used with any attribute because the function body itself is looking at the contents of specific parsed attribute kinds to decide what to do. So I'd say this function name should be left alone.

My hope is to eventually refactor this function either out of existence, or to make it a `static` function that's internal to SemaDeclAttr.cpp.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 2]
```diff
@@ -587,6 +587,11 @@ static void instantiateIntelFPGAMemoryAttr(
         std::is_same<AttrName, IntelFPGANumBanksAttr>::value)
       return S.AddOneConstantPowerTwoValueAttr<AttrName>(New, *Attr,
                                                          Result.getAs<Expr>());
+
+    else if (std::is_same<AttrName, IntelFPGAPrivateCopiesAttr>::value)
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:591]: No `else` after a `return`, so this can be an `if` rather than an `else if`.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 3]
```diff
@@ -629,15 +634,15 @@ static void instantiateSYCLIntelLoopFuseAttr(
 }
 
 template <typename AttrName>
-static void instantiateIntelSYCLFunctionAttr(
-    Sema &S, const MultiLevelTemplateArgumentList &TemplateArgs,
-    const AttrName *Attr, Decl *New) {
+static void
+instantiateFunctionAttr(Sema &S,
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:638]: Same concern about the rename here, tbh. This whole file is about template instantiation, and this function definitely does not instantiate generic function attributes (it's quite specific to intel's attributes).

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 4]
```diff
@@ -48,8 +48,6 @@ void check_ast()
   //CHECK: VarDecl{{.*}}private_copies
   //CHECK: IntelFPGAMemoryAttr{{.*}}Implicit
   //CHECK: IntelFPGAPrivateCopiesAttr
-  //CHECK-NEXT: ConstantExpr
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/intel-fpga-local.cpp:51]: This looks like a regression -- we're no longer retaining the attribute argument in the AST (so you can't AST match this value either). I think the reason why it's no longer showing up is because the given expression is no longer being converted into a constant expression -- we evaluate it as such, but then pass in the original expression (not the converted constant expression) when creating the attribute.

File: clang/test/SemaSYCL/scheduler_target_fmax_mhz.cpp [hunk 5]
```diff
@@ -10,20 +10,16 @@ template <int N>
 [[intel::scheduler_target_fmax_mhz(N)]] void zoo() {}
 
 int main() {
-  // CHECK-LABEL:  FunctionDecl {{.*}}test_kernel1 'void ()'
```
Comment (AaronBallman) [inline clang/test/SemaSYCL/scheduler_target_fmax_mhz.cpp:13]: Was this change intentional? `CHECK-LABEL` is used to "anchor" FileCheck's matching behavior, so when you remove the `-LABEL` part, it can start to match in surprising locations you weren't expecting. I don't see any changes that would necessitate dropping the `-LABEL` (here or below), so I'm kinda curious what's going on.

**Review by AaronBallman — COMMENTED** [review]
I think the rename is problematic due to how Intel-specific the logic of the functions is. I think we should at least keep `Intel` in the name so that it's clear this is highly specific to our needs. Ultimately, I think we should be very wary of any function that needs to look at the parsed attribute kind. That's usually a code smell that the function isn't particularly generic and its exposure should be limited as much as possible.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp (hunk 3 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:638]: Ok. With that in mind, would you prefer if I created separate functions i.e. `instantiatexyzattribute ` for each attribute, I am going to modify now? . It would result in a fair amount of code duplication but I believe this is how clang handles most attributes

File: clang/test/SemaSYCL/scheduler_target_fmax_mhz.cpp (hunk 5 above)
Comment (elizabethandrews) [reply clang/test/SemaSYCL/scheduler_target_fmax_mhz.cpp:13]: I thought it wasn't required, but I can add it back

File: clang/test/SemaSYCL/intel-fpga-local.cpp (hunk 4 above)
Comment (AaronBallman) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:51]: Good question! I'm thinking out loud here...

I could be wrong, but I'm pretty sure that `addIntelSYCLSingleArgFunctionAttr()` is calling `getIntegerConstantExpr()`, which goes through all the trouble of converting the expression into a constant expression but then throws the expr results away. So if we're going to go to the trouble of converting the expression to a constant expression, it seems reasonable to store the converted results if they're useful. AST dumping is an example of a case where it's somewhat useful (seeing the constant expression value is useful for humans, but more interestingly, there's a JSON dump that's used by other tools for automation and this would prevent those tools from having to do the work of evaluating the expression to get its result). AST matching also gets somewhat easier in that you can write an AST matcher that checks whether the argument has a specific value (but this is somewhat less compelling because AST matching on attribute arguments is really brand new functionality that's got a lot of rough edges). I think the most compelling reason is for ease of implementation within Clang itself -- if we convert the arguments to be a constant expression and store that in the semantic attribute, then users of the semantic attribute can rely on the expression having already been evaluated and more directly get the results without having to evaluate the constant expression a second time (such as when doing codegen).

So I think we should consistently store a `ConstantExpr` as the attribute argument, and then audit the places where we try to get the value out of the semantic attribute to see if there are places where we can remove duplicate effort (by calling `ConstantExpr::Int64Result()` or `ConstantExpr::APValueResult()` instead of re-evaluating the expression). WDYT?

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp (hunk 3 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:638]: I'd say hold off on that for now -- we may want to do it later, but for right now I think the logic is reasonable and the name is reasonable. Once we figure out a better way to split `addIntelSYCLSingleArgFunctionAttr()` up, we'll probably have a better idea of what to do here for template instantiation.

File: clang/test/SemaSYCL/scheduler_target_fmax_mhz.cpp (hunk 5 above)
Comment (AaronBallman) [reply clang/test/SemaSYCL/scheduler_target_fmax_mhz.cpp:13]: I don't think it's *required* (you can write a test without ever using it, certainly), but I think it makes the test a bit less fragile. I know I've been bitten by tests that accidentally passed because a `CHECK` line was finding something unexpected later in the file and using a `CHECK-LABEL` beforehand would have made the failure really obvious.

I hate FileCheck despite thinking it's pretty powerful. :-D

File: clang/test/SemaSYCL/intel-fpga-local.cpp (hunk 4 above)
Comment (elizabethandrews) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:51]: I think `AddOneConstantValueAttr` does the constant generation, not `addIntelSYCLSingleArgFunctionAttr `. I believe the difference lies in `VerifyIntegerConstantExpression `. This is called in `AddOneConstantValueAttr`. We're seeing it removed in this patch because I changed calls to `AddOneConstantValueAttr` with `addIntelSYCLSingleArgFunctionAttr ` because the formed has a call to `checkRangedIntegralArgument` which uses the additional members I'm removing in Attribute definition

Comment (elizabethandrews) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:51]: > So I think we should consistently store a `ConstantExpr` as the attribute argument, and then audit the places where we try to get the value out of the semantic attribute to see if there are places where we can remove duplicate effort (by calling `ConstantExpr::Int64Result()` or `ConstantExpr::APValueResult()` instead of re-evaluating the expression). WDYT?

Sounds good to me. I'll make the change. Thanks for the feedback!

Comment (AaronBallman) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:51]: I verified that my understanding was correct. `addIntelSYCLSingleArgFunctionAttr()` calls `E->getIntegerConstantExpr()` on the given expression, and `Expr::getIntegerConstantExpr()` evaluates the constant expression but throws away the expression results: https://github.com/intel/llvm/blob/sycl/clang/lib/AST/ExprConstant.cpp#L15518

Meanwhile, `AddOneConstantValueAttr()` calls `checkRangedIntegralArgument()` which calls `VerifyIntegerConstantExpression()` to convert the expression to a constant expression. This constant expression is returned to the caller, which resets the given expression to be the constant expression.

So either way, we're converting the expression to a constant expression, but only in one case are we throwing that useful work away (and I claim we shouldn't throw that work away but instead store it in the semantic attribute and make use of it elsewhere).

Comment (elizabethandrews) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:51]: Ok Thank you!

File: clang/include/clang/Sema/Sema.h [hunk 6]
```diff
@@ -13091,6 +13090,20 @@ void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
         return;
       }
     }
+    if (CI.getParsedKind() == ParsedAttr::AT_SYCLIntelSchedulerTargetFmaxMhz ||
+        CI.getParsedKind() == ParsedAttr::AT_IntelFPGAPrivateCopies) {
+      if (ArgInt < 0) {
+        Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+            << CI.getAttrName() << /*non-negative*/ 1;
```
Comment (AaronBallman) [inline clang/include/clang/Sema/Sema.h:13097]: You can drop the `getAttrName()` bit -- the diagnostics engine knows how to print a `CommonAttributeInfo` object properly already.

File: clang/include/clang/Sema/Sema.h [hunk 7]
```diff
@@ -13059,9 +13059,8 @@ class Sema final {
 };
 
 template <typename AttrType>
-void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
-                                             const AttributeCommonInfo &CI,
-                                             Expr *E) {
+void Sema::addIntelSingleArgFunctionAttr(Decl *D, const AttributeCommonInfo &CI,
```
Comment (elizabethandrews) [inline clang/include/clang/Sema/Sema.h:13062]: I renamed this to handle both IntelSYCL and IntelFPGA attributes in the same function. I In my opinion this function needs refactoring but I didn't do it in this patch because I think this overlaps with what @AaronBallman is working on.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:13062]: LGTM. One nit

Since you are adding memory attribute inside addIntelSingleArgFunctionAttr(), so it would be better if you remove "Function" from the function name here to handle both function and memory attributes.

Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h]: Thanks for catching that! Done

File: clang/include/clang/Sema/Sema.h (hunk 1 above)
Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h]: I removed the 'SYCL' and 'Function' from the name since I now use this function to handle other FPGA attributes as well. I'm not a fan of the new name `addIntelSingleArgAttr` but I'm not sure what else to call it. Please feel free to suggest another name or let me know if you'd like me to refactor this function in some way. I think the diagnostics should probably be moved out to the individual attribute handlers, but I think this work overlaps with the patch you're working on and so I left it as is.  

I think another look into this function, `AddOneConstantValueAttr `, `AddOneConstantPowerTwo `, `instantiateIntelFPGAMemoryAttr `and `instantiateIntelSYCLFunctionAttr` is required  at some point. All of these could possibly be refactored, maybe even removed.

Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h]: My long-term goal is to remove these functions that look like they should be generic but actually aren't generic. Ideally, we could use tablegen to generate code that turns out to be boilerplate in practice. So the name isn't super worrying right now by comparison to the function existing. :-D

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 8]
```diff
@@ -574,28 +574,37 @@ static void instantiateIntelSYCTripleLFunctionAttr(
...
-  }
+static void instantiateIntelFPGANumBanksAttr(
+    Sema &S, const MultiLevelTemplateArgumentList &TemplateArgs,
+    const IntelFPGANumBanksAttr *Attr, Decl *New) {
+  EnterExpressionEvaluationContext Unevaluated(
+      S, Sema::ExpressionEvaluationContext::ConstantEvaluated);
+  ExprResult Result = S.SubstExpr(Attr->getValue(), TemplateArgs);
+  if (!Result.isInvalid())
+    S.AddOneConstantPowerTwoValueAttr<IntelFPGANumBanksAttr>(
+        New, *Attr, Result.getAs<Expr>());
```
Comment (smanna12) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp]: both are doing same. You can combine them to one. Is there a reason to have a separate instantiation function.

File: clang/include/clang/Sema/Sema.h [hunk 9]
```diff
@@ -13164,13 +13175,6 @@ void Sema::AddOneConstantValueAttr(Decl *D, const AttributeCommonInfo &CI,
       return;
     E = ICE.get();
   }
-
-  if (IntelFPGAPrivateCopiesAttr::classof(&TmpAttr)) {
```
Comment (smanna12) [inline clang/include/clang/Sema/Sema.h:13168]: We do not use AddOneConstantValueAttr() any more. I think we can remove this redundant function.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp (hunk 8 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp]: I noticed open-source clang usually kept it separate and so I did as well.

Comment (AaronBallman) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp]: Please keep them separate for now as I'm expecting that to make future work a bit easier.

File: clang/include/clang/Sema/Sema.h (hunk 9 above)
Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h:13168]: It is still used when handling `IntelFPGAForcePow2DepthAttr`. I can rename it to `addIntelFPGAForcePow2Depth()` if you'd like. I wasn't sure if other FPGA attributes (For example, other attributes with ranges) could/would be refactored to use this function in the future, and so I left it as it is. Should I rename it?

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:13168]: Thanks for the clarification. I did not realize that It is still used when handling IntelFPGAForcePow2DepthAttr. Rename is not needed here.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp (hunk 8 above)
Comment (smanna12) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp]: > I noticed open-source clang usually kept it separate and so I did as well.

Thanks for the info and clarification.

File: clang/include/clang/Sema/Sema.h [hunk 10]
```diff
@@ -13072,27 +13069,42 @@ void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
...
+    }
+    if (CI.getParsedKind() == ParsedAttr::AT_SYCLIntelSchedulerTargetFmaxMhz ||
+        CI.getParsedKind() == ParsedAttr::AT_IntelFPGAPrivateCopies) {
+      if (ArgInt < 0) {
+        Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+            << CI << /*non-negative*/ 1;
         return;
       }
     }
   }
```
Comment (smanna12) [inline clang/include/clang/Sema/Sema.h:13101]: Is there a reason you have added check (ArgInt < 0) again for SYCLIntelSchedulerTargetFmaxMhzAttr and IntelntelFPGAPrivateCopiesAttr here?

This is same as lines +13080 to + 13085. You can combine like
CI.getParsedKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim ||
CI.getParsedKind() == ParsedAttr::AT_SYCLIntelSchedulerTargetFmaxMhz ||
CI.getParsedKind() == ParsedAttr::AT_IntelFPGAPrivateCopies

Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h:13101]: SYCLIntelMaxGlobalWorkDim has an upper bound check on Line 13086, which is not required for SYCLIntelSchedulerTargetFmaxMhz  and IntelFPGAPrivateCopies. Instead of creating additional nested ifs, I kept the bound checks for SYCLIntelSchedulerTargetFmaxMhz and IntelFPGAPrivateCopies since the requirements are different. It felt cleaner this way.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:13101]: Yup, right.

**Review by smanna12 — APPROVED** [review]

File: clang/test/SemaSYCL/loop_fusion.cpp [hunk 11]
```diff
@@ -4,8 +4,8 @@
 
 [[intel::loop_fuse("foo")]] void func() {} // expected-error{{'loop_fuse' attribute requires an integer constant}}
 
-[[intel::loop_fuse(1048577)]] void func1() {}        // expected-error{{'loop_fuse' attribute requires integer constant between 0 and 1048576 inclusive}}
-[[intel::loop_fuse_independent(-1)]] void func2() {} // expected-error{{'loop_fuse_independent' attribute requires integer constant between 0 and 1048576 inclusive}}
+[[intel::loop_fuse(1048577)]] void func1() {}        // OK
```
Comment (premanandrao) [inline clang/test/SemaSYCL/loop_fusion.cpp:7]: I am okay with the changes.   Just a question about the upper bound.   Is there an upper bound at all now?

Comment (elizabethandrews) [reply clang/test/SemaSYCL/loop_fusion.cpp:7]: For these attributes, there isn't. As discussed in https://github.com/intel/llvm/issues/2911, there is no need for an upper bound since the backend doesn't care.

**Review by premanandrao — APPROVED** [review]

**Comment by elizabethandrews** [comment]
Thanks for the reviews everyone!
